doc/elasticdl_on_sqlflow.md

```diff
@@ -0,0 +1,141 @@
+# _Design:_ ElasticDL on sqlflow
+
+## Overview
+
+This is a design doc on integration with [ElasticDL](https://github.com/wangkuiyi/elasticdl).
+
+### User Interface
+
+#### Training Job Submission
+
+```sql
+SELECT 
+    c1, c2, c3, c4, c5 as class
+FROM training_data
+TRAIN ElasticDLKerasClassifier
+WITH
+  optimizer = "optimizer"
+  loss = "loss"
+  eval_metrics = "eval_metrics_fn"
+  num_classes = 10
+COLUMN
+  c1,
+  NUMERIC(c2, 10),
+  BUCKET(c3, [0, 10, 100]),
+  c4
+LABEL class
+INTO trained_elasticdl_keras_classifier;
+```
+
+#### Prediction Job Submission
+
+```sql
+SELECT 
+    c1, c2, c3, c4
+FROM prediction_data
+PREDICT prediction_results_table
+WITH
+  num_classes = 10
+USING trained_elasticdl_keras_classifier;
+```
+
+#### Run-time Configurations
+
+Users can provide run-time configurations to ElasticDL job via additional parameters with prefix "runtime" within `WITH` clause, for example:
+
+```sql
+SELECT
+    c1, c2, c3, c4, c5 as class
+FROM training_data
+TRAIN ElasticDLKerasClassifier
+WITH
+  optimizer = "optimizer"
+  loss = "loss"
+  eval_metrics = "eval_metrics_fn"
+  num_classes = 10
+  runtime.num_epochs = 2
+  runtime.master_resource_request = "cpu=400m,memory=1024Mi"
+  runtime.master_resource_limit = "cpu=400m,memory=1024Mi"
+  runtime.worker_resource_request = "cpu=400m,memory=2048Mi"
+  runtime.worker_resource_limit = "cpu=1,memory=3072Mi"
+  runtime.records_per_task = 100
+  runtime.num_workers = 2
+COLUMN
+  c1, c2, c3, c4
+LABEL class
+INTO trained_elasticdl_keras_classifier;
+```
+
+### Implementation Details
+
+#### Training Job
+
+Steps:
+
+1. Based on `SELECT ... FROM ...`, read ODPS table and write it to [RecordIO](https://github.com/wangkuiyi/recordio) files, including both features and labels.
+2. Generate model definition file (e.g. [cifar10_functional_api.py](https://github.com/wangkuiyi/elasticdl/blob/develop/model_zoo/cifar10_functional_api/cifar10_functional_api.py)) that will be used in `TRAIN` clause, which includes:
+    * In model definition function e.g. `custom_model()`, we need to configure model input and output shapes correctly in `inputs = tf.keras.layers.Input(shape=<input_shape>)` and `outputs = tf.keras.layers.Dense(<num_classes>)`(based on `COLUMN ... LABEL ...`). For this MVP, users can provide `<input_shape>` and `<num_classes>` using `WITH` clause which will then get passed to the model constructor `custom_model(input_shape, num_classes)` via `--params` argument in ElasticDL high-level API. In the future, this will be inferred from the ODPS table.
```

Comment:
SQLFlow can help with inferring data types and tensor shapes. :)

Reply:
Great! Is there an example for that?

Reply:
For Keras functional API, input layers are required. If the Keras subclass is used, no need to define the input layer.

Reply:
> Great! Is there an example for that?

@terrytangyuan the data type inference logic is scattered around in the code. I apologize for that.

For example, to determine the label type, we have

https://github.com/sql-machine-learning/sqlflow/blob/d873bc994350493a599e58a7cc5dc519d09e58e5/sql/codegen.go#L190-L205

To determine the input tensor shape in the case of dense tensor like "1,2,3,4". We select one row and split the cell by delimiter to determine the shape.

https://github.com/sql-machine-learning/sqlflow/blob/d873bc994350493a599e58a7cc5dc519d09e58e5/sql/codegen_alps.go#L658-L661

Currently, the type and shape inference logic is not unified across different `codegen_*.go`. SQLFlow team is working on unifying them and put the inferred result into a struct like`featureMeta`.

https://github.com/sql-machine-learning/sqlflow/blob/d873bc994350493a599e58a7cc5dc519d09e58e5/sql/codegen.go#L47-L53

This effort may take several weeks.

What is the first use case of EDL on SQLFlow?